python/tank/context.py

```diff
@@ -737,10 +737,133 @@ def from_entity(tk, entity_type, entity_id):
         
         context.update(entity_context)
 
+        if entity_type == "Project":
+            # no need to set entity to point at project in this case
+            # that only produces double entries.
+            context["entity"] = None
+
+    return Context(**context)
+
+def from_entity_dictionary(tk, entity_dictionary):
+    """
+    Constructs a context from a shotgun entity dictionary.  This method will
+    try to use any linked information available in the dictionary where possible
+    but if it can't determine a valid context then it will fall back to
+    'from_entity' which may result in a Shotgun path cache query and be considerably
+    slower.
```

Comment:
I think it would be cool to mention at some point that this is a recursive method. In my experience, this is a minor, however often an important piece of information for code readers and maintainers and useful to know when doing debugging.